DasharoPayloadPkg/Library/PlatformBootManagerLib/PlatformBootManager.c

```diff
@@ -1067,6 +1070,221 @@ WarnIfRecoveryBoot (
+  UINT32 Attr  = EFI_VARIABLE_NON_VOLATILE | EFI_VARIABLE_BOOTSERVICE_ACCESS | EFI_VARIABLE_RUNTIME_ACCESS;
+  EFI_STATUS Status = gRT->GetVariable(
+                    L"RequestedActivePcrBanks",
+                    &gEfiTcg2PhysicalPresenceGuid,
+                    NULL,
+                    &Size,
+                    &RequestedActivePcrBanks
+                  );
+  if (Status == EFI_NOT_FOUND) {
+    gRT->SetVariable(
+      REQUESTED_ACTIVE_PCR_BANKS_VARIABLE_NAME,
+      &gEfiTcg2PhysicalPresenceGuid,
+      Attr,
+      sizeof(RequestedActivePcrBanks),
+      &RequestedActivePcrBanks
+    );
+    DEBUG ((DEBUG_INFO, "RequestedActivePcrBanks variable set\n"));
+  } else {
+    DEBUG ((DEBUG_INFO, "RequestedActivePcrBanks variable already present, value: %x\n", RequestedActivePcrBanks));
+  }
+}
+
+
+typedef struct {
+  TPM_ALG_ID AlgId;
+  CHAR16    *Name;
+} HASH_ALG_ENTRY;
+
+STATIC CONST HASH_ALG_ENTRY mHashAlgs[] = {
+  { TPM_ALG_SHA1,    L"SHA1"    },
+  { TPM_ALG_SHA256,  L"SHA256"  },
+  { TPM_ALG_SHA384,  L"SHA384"  },
+  { TPM_ALG_SHA512,  L"SHA512"  },
+  { TPM_ALG_SM3_256, L"SM3_256" },
+};
+
+STATIC
+VOID
+WarnIfSinglePCRBank (
+  VOID
+)
+{
+  EFI_STATUS     Status;
+  EFI_EVENT      TimerEvent;
+  EFI_EVENT      Events[2];
+  BOOLEAN        CursorVisible;
+  UINTN          CurrentAttribute;
+  UINTN          Index;
+  EFI_INPUT_KEY  Key;
+  UINT32         TpmHashAlgorithmBitmap;
+  UINT32         ActivePcrBanks;
+  UINT32         RequestedActivePcrBanks;
+  UINTN          Size;
+  UINTN          i;
+  UINTN          OptionCount = 0;
+  UINTN          AvlIdx[ARRAY_SIZE(mHashAlgs)] = {0};
+  CHAR16         OptLine[81];
+  UINT32         SelectedMask;
+
+  Status = Tpm2GetCapabilitySupportedAndActivePcrs(&TpmHashAlgorithmBitmap, &ActivePcrBanks);
+  ASSERT_EFI_ERROR(Status);
+  Size = sizeof(RequestedActivePcrBanks);
+  Status = gRT->GetVariable(
+            L"RequestedActivePcrBanks",
+            &gEfiTcg2PhysicalPresenceGuid,
+            NULL,
+            &Size,
+            &RequestedActivePcrBanks
+            );
+  // It's zero by default, that means there's been no request to change the PCR banks.
+  if (RequestedActivePcrBanks == 0) {
+    return;
+  } else if (RequestedActivePcrBanks == ActivePcrBanks) {
+    // 
+    // If they're equal, zero it out again and proceed with normal boot.
+    // It means the requested bank selection was successful.
+    // 
+    RequestedActivePcrBanks = 0;
+    gRT->SetVariable(
+          L"RequestedActivePcrBanks",
+          &gEfiTcg2PhysicalPresenceGuid,
+          EFI_VARIABLE_NON_VOLATILE | EFI_VARIABLE_BOOTSERVICE_ACCESS | EFI_VARIABLE_RUNTIME_ACCESS,
+          sizeof(RequestedActivePcrBanks),
+          &RequestedActivePcrBanks
+        );
+    return;
+  }
+  // 
+  // If they're not equal, display the popup and switch to a single bank
+  // of user's choice.
+  // 
+  Status = gBS->CreateEvent(EVT_TIMER, TPL_CALLBACK, NULL, NULL, &TimerEvent);

```

Comment:
You're not doing anything when the timer is signaled, so no need to even create one.